### PR TITLE
Fix https://github.com/nrc/rustfmt/issues/270

### PR DESCRIPTION
**Pull request by mwiczer**
The variants of an enum use the list module rather than its own special case handling. This makes for better pre- and post-comment handling.

Additionally, there was a small bug in struct formatting that I never reported but resolved in this PR. I'll document it here:
If a field had any attributes and the post-comment needed to be split into a multi-line, the subsequent comment lines were misaligned. e.g.:

````rust
struct FieldsWithAttributes {
    // Pre Comment
    #[rustfmt_skip] pub host:String, /* Post comment BBBBBBBBBBBBBB BBBBBBBBBBBBBBBB BBBBBBBBBBBBBBBB
                                      * BBBBBBBBBBBBBBBBB BBBBBBBBBBB */
    // Another pre comment
    #[attr1]
    #[attr2]
    pub id: usize, /* CCCCCCCCCCCCCCCCCCC CCCCCCCCCCCCCCCCCCC CCCCCCCCCCCCCCCC CCCCCCCCCCCCCCCCCC
                    * CCCCCCCCCCCCCC CCCCCCCCCCCC */
}
````

was formatted to: 
````rust
struct FieldsWithAttributes {
    // Pre Comment
    #[rustfmt_skip] pub host:String, /* Post comment BBBBBBBBBBBBBB BBBBBBBBBBBBBBBB BBBBBBBBBBBBBBBB
                                      * BBBBBBBBBBBBBBBBB BBBBBBBBBBB */
    // Another pre comment
    #[attr1]
    #[attr2]
    pub id: usize, /* CCCCCCCCCCCCCCCCCCC CCCCCCCCCCCCCCCCCCC
                                              * CCCCCCCCCCCCCCCC CCCCCCCCCCCCCCCCCC
                                              * CCCCCCCCCCCCCC CCCCCCCCCCCC */
}
````

### CODE REVIEW

File: src/items.rs [hunk 1]
```diff
@@ -592,31 +592,95 @@ impl<'a> FmtVisitor<'a> {
...
+
+        if contains_skip(&field.node.attrs) {
+
+            let lo = field.node.attrs[0].span.lo;
+            let span = codemap::mk_sp(lo, field.span.hi);
+            return Some(self.snippet(span));
+        }
+
+        let indent = self.block_indent;
+        let mut result = field.node
```
Comment (marcusklaas) [inline src/items.rs]: We should wrap this in `try_opt!`, so we fail gracefully instead of panicking.

File: src/items.rs [hunk 2]
```diff
@@ -592,31 +592,95 @@ impl<'a> FmtVisitor<'a> {
...
+            tactic: DefinitiveListTactic::Vertical,
+            separator: ",",
+            trailing_separator: SeparatorTactic::Always,
+            indent: self.block_indent,
+            width: budget,
+            ends_with_newline: true,
+            config: self.config,
+        };
+
+        let mut result = try_opt!(write_list(&items, &fmt));
```
Comment (marcusklaas) [inline src/items.rs]: We can pass the iterator directly here. No need to allocate a vector, woo!

File: src/items.rs [hunk 3]
```diff
@@ -592,31 +592,95 @@ impl<'a> FmtVisitor<'a> {
...
+        self.last_pos = span.hi;
+    }
+
+    // Format the body of an enum definition
+    fn format_variant_list(&self,
+                           enum_def: &ast::EnumDef,
+                           body_lo: BytePos,
+                           body_hi: BytePos)
+                           -> Option<String> {
+
```
Comment (marcusklaas) [inline src/items.rs]: Nit: the blank lines after opening braces are not really in line with the rest of the code.

**Comment by nrc** [comment]
Could you tidy up the git history before landing please? Probably the easiest way to do this is to pull master, checkout a fresh branch and then cherry pick your commits. Thanks!

**Comment by marcusklaas** [comment]
:100: 

These are great changes!

**Comment by mwiczer** [comment]
Thanks for the help, guys. I've made a new PR with a cleaned-up history -- #420. 

I explicitly marked the latest changes I made because your comments here won't migrate. Hope it wasn't annoyingly verbose of me `emoticon`.

**Pull request closed without merging**